File: native/expo-zechinus-crypto/android/src/main/java/expo/modules/zechinuscrypto/CryptoKeyRef.kt
package expo.modules.zechinuscrypto

import android.util.Base64
import com.google.crypto.tink.subtle.Hkdf
import expo.modules.kotlin.sharedobjects.SharedObject
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

class ZechinusCryptoException(message: String) : Exception(message)

/**
 * One key, one instance — see `CryptoKeyRef.swift` for the rationale. Key material is
 * never returned to JS by any method here.
 *
 * `android.util.Base64`, NOT `java.util.Base64` — the latter requires API 26 and would
 * crash on the older devices Expo's minSdk still allows. `NO_WRAP` is mandatory: the
 * default inserts newlines and breaks interop.
 */
class CryptoKeyRef(rawBytes: ByteArray) : SharedObject() {
  private var key: ByteArray?
  private var macKey: ByteArray? = null

  init {
    if (rawBytes.size != 32) {
      throw ZechinusCryptoException("CryptoKey: key must be 32 bytes, got ${rawBytes.size}")
    }
    key = rawBytes.copyOf()
  }

  private fun requireKey(): ByteArray = key ?: throw ZechinusCryptoException("CryptoKey: already destroyed")

  fun initMacKey(bytes: ByteArray) {
    requireKey()
    macKey = bytes.copyOf()
  }

  fun aesGcmEncrypt(nonce: ByteArray, aad: ByteArray, plaintext: ByteArray): ByteArray {
    val k = requireKey()
    if (nonce.size != 12) throw ZechinusCryptoException("CryptoKey: nonce must be 12 bytes, got ${nonce.size}")
    val cipher = Cipher.getInstance("AES/GCM/NoPadding")
    cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(k, "AES"), GCMParameterSpec(128, nonce))
    cipher.updateAAD(aad)
    return cipher.doFinal(plaintext) // ciphertext||tag, same as @noble/ciphers
  }

  fun aesGcmDecrypt(nonce: ByteArray, aad: ByteArray, ciphertext: ByteArray): ByteArray {
    val k = requireKey()
    if (nonce.size != 12) throw ZechinusCryptoException("CryptoKey: nonce must be 12 bytes, got ${nonce.size}")
    if (ciphertext.size <= 16) {
      throw ZechinusCryptoException(
        "CryptoKey: ciphertext must include the 16-byte GCM tag, got ${ciphertext.size} bytes"
      )
    }
    val cipher = Cipher.getInstance("AES/GCM/NoPadding")
    cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(k, "AES"), GCMParameterSpec(128, nonce))
    cipher.updateAAD(aad)
    return cipher.doFinal(ciphertext)
  }

  /**
   * HKDF-SHA256 via Tink, NOT hand-written. AES-GCM and HMAC come from the platform;
   * HKDF is the one primitive Android's `javax.crypto` does not expose under a standard
   * algorithm name, so the choice is a maintained library or our own code. Fifteen lines
   * of hand-written key derivation is not a good trade for one dependency.
   */
  fun hkdfDerive(salt: ByteArray, info: String, length: Int): ByteArray =
    Hkdf.computeHkdf("HMACSHA256", requireKey(), salt, info.toByteArray(Charsets.UTF_8), length)

  fun hmacSha256(payload: ByteArray): String {
    requireKey()
    val mk = macKey ?: throw ZechinusCryptoException("CryptoKey: no MAC key — call initMacKey first")
    val mac = Mac.getInstance("HmacSHA256")
    mac.init(SecretKeySpec(mk, "HmacSHA256"))
    return mac.doFinal(payload).joinToString("") { "%02x".format(it) }
  }

  fun wrapSelf(kek: ByteArray): Map<String, String> {
    val k = requireKey()
    val nonce = ByteArray(12).also { SecureRandom().nextBytes(it) }
    val cipher = Cipher.getInstance("AES/GCM/NoPadding")
    cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(kek, "AES"), GCMParameterSpec(128, nonce))
    return mapOf(
      "ciphertext" to Base64.encodeToString(cipher.doFinal(k), Base64.NO_WRAP),
      "nonce" to Base64.encodeToString(nonce, Base64.NO_WRAP),
    )
  }

  /** Idempotent. Zeroing before dropping the reference is the point — the JVM won't do it. */
  fun destroy() {
    key?.fill(0)
    macKey?.fill(0)
    key = null
    macKey = null
  }

  /**
   * The only place in this module that turns key material back into plain bytes —
   * `internal` (not exposed via the `Class("CryptoKey", ...)` binding), so it exists for
   * `ZeroTapKeystoreStore` in this same module to call, and is unreachable from JS: it is
   * never listed inside the `Class` definition, so the Expo Modules bridge has no way to
   * invoke it. The returned hex string goes straight into an AndroidKeyStore-encrypted
   * blob gated by `BiometricPrompt` (`ZeroTapKeystoreStore.cache`) — it is never logged,
   * returned across the JS bridge, or persisted anywhere else.
   */
  internal fun exportForKeychainOnly(): String =
    requireKey().joinToString("") { "%02x".format(it) }
}

File: native/expo-zechinus-crypto/ios/CryptoKeyRef.swift
import CryptoKit
import ExpoModulesCore
import Foundation

enum ZechinusCryptoError: Error, LocalizedError {
  case destroyed
  case noMacKey
  case invalidKeyLength(Int)
  case invalidNonceLength(Int)
  case ciphertextTooShort(Int)
  case invalidHkdfLength(Int)
  case accessControlUnavailable
  case keychainWrite(OSStatus)

  var errorDescription: String? {
    switch self {
    case .destroyed: return "CryptoKey: already destroyed"
    case .noMacKey: return "CryptoKey: no MAC key — call initMacKey first"
    case .invalidKeyLength(let n): return "CryptoKey: key must be 32 bytes, got \(n)"
    case .invalidNonceLength(let n): return "CryptoKey: nonce must be 12 bytes, got \(n)"
    case .ciphertextTooShort(let n): return "CryptoKey: ciphertext must include the 16-byte GCM tag, got \(n) bytes"
    case .invalidHkdfLength(let n): return "CryptoKey: HKDF length must be 1...8160, got \(n)"
    case .accessControlUnavailable: return "CryptoKey: could not create a biometric-gated Keychain access control (no biometry enrolled on this device?)"
    case .keychainWrite(let status): return "CryptoKey: Keychain write failed with status \(status)"
    }
  }
}

/// What `cacheKeyForZeroTap` persists to the Keychain, opaque to anything reading it
/// outside this module — the biometric ACL on the Keychain item is what actually
/// protects it, this struct only describes the shape of the plaintext-once-inside-
/// Secure-Enclave-access payload.
struct CachedEntry: Codable {
  let keyHex: String
  let dekEpoch: Int
  let credentialId: String
}

extension Data {
  /// Lowercase-or-uppercase hex string -> bytes. `nil` on malformed input (odd length,
  /// non-hex characters) rather than trapping — a corrupted/tampered Keychain entry
  /// must fail closed (`tryRestoreFromNativeCache` returns `nil`), never crash.
  init?(hexString: String) {
    guard hexString.count % 2 == 0 else { return nil }
    var bytes = [UInt8]()
    bytes.reserveCapacity(hexString.count / 2)
    var index = hexString.startIndex
    while index < hexString.endIndex {
      let next = hexString.index(index, offsetBy: 2)
      guard let byte = UInt8(hexString[index..<next], radix: 16) else { return nil }
      bytes.append(byte)
      index = next
    }
    self = Data(bytes)
  }

  var hexString: String {
    map { String(format: "%02x", $0) }.joined()
  }
}

/// One key, one instance — the native counterpart of one Web Worker on web. Two handles
/// alive at once (a DEK rotation keeps `previousCryptoHandle` alongside the current one)
/// means two independent instances, so `destroy()` on one cannot reach the other.
/// The key material is never returned to JS by any method here.
public final class CryptoKeyRef: SharedObject {
  private var key: SymmetricKey?
  private var macKey: SymmetricKey?

  public init(rawBytes: Data) throws {
    guard rawBytes.count == 32 else { throw ZechinusCryptoError.invalidKeyLength(rawBytes.count) }
    self.key = SymmetricKey(data: rawBytes)
    super.init()
  }

  private func requireKey() throws -> SymmetricKey {
    guard let key else { throw ZechinusCryptoError.destroyed }
    return key
  }

  public func initMacKey(_ bytes: Data) throws {
    _ = try requireKey()
    macKey = SymmetricKey(data: bytes)
  }

  public func aesGcmEncrypt(_ nonce: Data, _ aad: Data, _ plaintext: Data) throws -> Data {
    let key = try requireKey()
    guard nonce.count == 12 else { throw ZechinusCryptoError.invalidNonceLength(nonce.count) }
    let sealed = try AES.GCM.seal(
      plaintext, using: key, nonce: try AES.GCM.Nonce(data: nonce), authenticating: aad
    )
    // Convention matching @noble/ciphers: ciphertext with the 16-byte GCM tag appended.
    return sealed.ciphertext + sealed.tag
  }

  public func aesGcmDecrypt(_ nonce: Data, _ aad: Data, _ ciphertext: Data) throws -> Data {
    let key = try requireKey()
    guard nonce.count == 12 else { throw ZechinusCryptoError.invalidNonceLength(nonce.count) }
    guard ciphertext.count > 16 else { throw ZechinusCryptoError.ciphertextTooShort(ciphertext.count) }
    // `Data(...)` around each slice is REQUIRED: a Data slice keeps the parent's index
    // base, so slicing an already-sliced Data reads from the wrong offset. Re-wrapping
    // rebases to 0. Task A6 exercises this path through this method, not through raw
    // CryptoKit — which is the only way the bug would have been caught.
    let body = Data(ciphertext.prefix(ciphertext.count - 16))
    let tag = Data(ciphertext.suffix(16))
    let sealed = try AES.GCM.SealedBox(
      nonce: try AES.GCM.Nonce(data: nonce), ciphertext: body, tag: tag
    )
    return try AES.GCM.open(sealed, using: key, authenticating: aad)
  }

  public func hkdfDerive(_ salt: Data, _ info: String, _ length: Int) throws -> Data {
    let key = try requireKey()
    guard length > 0 && length <= 255 * 32 else { throw ZechinusCryptoError.invalidHkdfLength(length) }
    let derived = HKDF<SHA256>.deriveKey(
      inputKeyMaterial: key, salt: salt, info: Data(info.utf8), outputByteCount: length
    )
    return derived.withUnsafeBytes { Data($0) }
  }

  public func hmacSha256(_ payload: Data) throws -> String {
    _ = try requireKey()
    guard let macKey else { throw ZechinusCryptoError.noMacKey }
    return Data(HMAC<SHA256>.authenticationCode(for: payload, using: macKey))
      .map { String(format: "%02x", $0) }.joined()
  }

  /// Wraps THIS key (not one passed in) under the given KEK. No explicit nonce:
  /// `AES.GCM.seal` generates a fresh cryptographically secure one.
  public func wrapSelf(_ kek: Data) throws -> [String: String] {
    let key = try requireKey()
    let sealed = try AES.GCM.seal(key.withUnsafeBytes { Data($0) }, using: SymmetricKey(data: kek))
    return [
      "ciphertext": (sealed.ciphertext + sealed.tag).base64EncodedString(),
      "nonce": Data(sealed.nonce).base64EncodedString(),
    ]
  }

  /// Idempotent: destroying twice is legitimate. Dropping the reference is enough —
  /// CryptoKit's SymmetricKey zeroes its own backing store on deallocation.
  public func destroy() {
    key = nil
    macKey = nil
  }

  /// The only place in this module that turns key material back into plain bytes —
  /// `internal` (not `public`), so it exists for `ExpoZechinusCryptoModule`'s Keychain
  /// functions in this same target to call, and is unreachable from JS: it is never
  /// listed inside `Class(CryptoKey.self) { ... }`, so the Expo Modules bridge has no
  /// way to invoke it. The exported hex string goes straight into a Keychain item
  /// gated by `.biometryCurrentSet` (`cacheKeyForZeroTap`) — it is never logged,
  /// returned across the JS bridge, or persisted anywhere else.
  func exportForKeychainOnly() throws -> String {
    try requireKey().withUnsafeBytes { Data($0) }.hexString
  }
}
